README.md
# app_upgrade

*Android download and upgrade APK, custom UI components support download progress callback*

## AndroidManifest.xml application 增加
```xml
 <provider
        android:name="androidx.core.content.FileProvider"
        android:authorities="${applicationId}.fileprovider"
        android:exported="false"
        android:grantUriPermissions="true">
        <meta-data
            android:name="android.support.FILE_PROVIDER_PATHS"
            tools:replace="android:resource"
            android:resource="@xml/file_paths"/>
 </provider>

```
## 添加依赖
```yaml
dependencies:
  sr_upgrade: ^0.0.2
```

## 使用
- 引入
```dart
import 'package:flutter/material.dart';
import 'package:sr_upgrade/sr_upgrade.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  @override
  void initState() {
    super.initState();
  }

  // Platform messages are asynchronous, so we initialize in an async method.

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Home(),
    );
  }
}

class Home extends StatefulWidget {
  @override
  _HomeState createState() => _HomeState();
}

class _HomeState extends State<Home> {
  String step = '0.00';

  _upgrade() {
    SrUpgrade.upgrade(
        'https://dldir1.qq.com/weixin/android/weixin806android1900_arm64.apk',
        progressCallback: (int count, int total) {
      setState(() {
        step = (count / total * 100).toStringAsFixed(2);
      });
    });
  }


  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Plugin example app'),
      ),
      body: Container(
        width: double.infinity,
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          crossAxisAlignment: CrossAxisAlignment.center,
          children: [
            TextButton(
              onPressed: _upgrade,
              child: Container(
                width: 200,
                height: 50,
                decoration: BoxDecoration(
                    color: Colors.blue,
                    borderRadius: BorderRadius.circular(15)),
                alignment: Alignment.center,
                child: Text(
                  '检查升级',
                  style: TextStyle(color: Colors.white, fontSize: 20),
                ),
              ),
            ),
            Text('当前下载进度$step%')
          ],
        ),
      ),
    );
  }
}

```


CHANGELOG.md
## 0.0.1

* TODO: Describe initial release.

## 0.0.2

* TODO: remove dialog UI


test/app_upgrade_test.dart
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  const MethodChannel channel = MethodChannel('app_upgrade');

  TestWidgetsFlutterBinding.ensureInitialized();

  setUp(() {
    channel.setMockMethodCallHandler((MethodCall methodCall) async {
      return '42';
    });
  });

  tearDown(() {
    channel.setMockMethodCallHandler(null);
  });

  test('getPlatformVersion', () async {

  });
}


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:app_upgrade_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) => widget is Text &&
                           widget.data!.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


lib/sr_upgrade.dart
import 'package:dio/dio.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

typedef ProgressCallback = void Function(int count, int total);
typedef DioCallback = void Function(Dio dio);

///[getPath] 获取apk存储路径
///[upgrade] 升级
///[installApk] 安装apk
class SrUpgrade {
  static const MethodChannel _channel = const MethodChannel('app_upgrade');

  static getPath() async {
    return await _channel.invokeMethod('path');
  }

  static Future upgrade<bool>(String downloadUrl,
      {ProgressCallback? progressCallback, DioCallback? dioCallback}) async {
    final String rootPath = await getPath();
    final String appPath = '$rootPath/temp_apk.apk';
    final dio = Dio();
    CancelToken token = new CancelToken();
    try {
      await dio.download(downloadUrl, appPath,cancelToken: token,
          deleteOnError: true,
          onReceiveProgress: progressCallback ?? (int count, int total) {});
      if (dioCallback != null) {
        dioCallback(dio);
      }
    } catch (err) {
      return false;
    }
    return await installApk(appPath);
  }

  static Future installApk<bool>(String path) async {
    try {
      await _channel.invokeMethod('install', {"path": path});
    } catch (err) {
      return false;
    }
    return true;
  }
}


lib/app_upgrade.dart
import 'package:app_upgrade/dialog.dart';
import 'package:app_upgrade/modal.dart';
import 'package:dio/dio.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

import 'default_dialog.dart';
export 'modal.dart';

typedef ProgressCallback = void Function(int count, int total);

///[show] 显示升级弹窗
///[getPath] 获取apk存储路径
///[upgrade] 升级
///[installApk] 安装apk
class AppUpgrade {
  static const MethodChannel _channel = const MethodChannel('app_upgrade');

  static late BuildContext context;

  static ProgressCallback? onProgressCallback;
  static late bool isDownloadCloseForUpgradeWindow;

  ///[context] buildContent
  ///[info] 升级弹窗信息model[AppUpgradeModel]
  ///[progressCallback] 下载进度回调函数[ProgressCallback]
  ///[AppUpgradeDialog] 自定义弹窗组件需要继承实现onCancel onOk 方法
  ///[downloadCloseForUpgradeWindow] 下载的时候是否关闭弹窗 默认为false
  static show(BuildContext context,
      {required AppUpgradeModel info,
      AppUpgradeDialog? appUpgradeDialog,
      ProgressCallback? progressCallback,
      bool? downloadCloseForUpgradeWindow = false}) {
    AppUpgrade.context = context;
    onProgressCallback = progressCallback;
    isDownloadCloseForUpgradeWindow = downloadCloseForUpgradeWindow ?? false;
    showDialog(
        context: context,
        barrierDismissible: false,
        builder: (BuildContext context) =>
            appUpgradeDialog ??
            AppUpgradeDialogDefault(
              info: info,
            ));
  }

  static getPath() async {
    return await _channel.invokeMethod('path');
  }

  static Future upgrade<bool>(String downloadUrl) async {
    // ignore: unnecessary_null_comparison
    if (context == null || isDownloadCloseForUpgradeWindow == null) {
      throw UnimplementedError('请先调用show方法');
    }
    final String rootPath = await getPath();
    final String appPath = '$rootPath/temp_apk.apk';
    final dio = Dio();
    if (isDownloadCloseForUpgradeWindow) {
      Navigator.pop(context);
    }
    try {
      await dio.download(downloadUrl, appPath,
          deleteOnError: true, onReceiveProgress: onProgressCallback);
    } catch (err) {
      return false;
    }
    if (!isDownloadCloseForUpgradeWindow) {
      Navigator.pop(context);
    }
    return await installApk(appPath);
  }

  static Future installApk<bool>(String path) async {
    try {
      await _channel.invokeMethod('install', {"path": path});
    } catch (err) {
      return false;
    }
    return true;
  }
}


lib/default_dialog.dart
import 'package:app_upgrade/dialog.dart';
import 'package:app_upgrade/modal.dart';
import 'package:flutter/material.dart';

import 'app_upgrade.dart';

class AppUpgradeDialogDefault extends AppUpgradeDialog {
  final AppUpgradeModel info;

  AppUpgradeDialogDefault({required this.info});

  @override
  Widget build(BuildContext context) {
    return Material(
      child: Container(
        alignment: Alignment.center,
        child: Container(
          width: 250,
          height: 300,
          decoration: BoxDecoration(
            color: Colors.white,
            borderRadius: BorderRadius.circular(5),
          ),
          child: Column(
            children: [
              Text(
                info.title ?? '',
                style: TextStyle(
                    fontSize: 16, color: Color(0xff333333), height: 2),
              ),
              Expanded(
                  child: ListView(
                padding: EdgeInsets.symmetric(horizontal: 15),
                children: info.contents!
                    .map((e) => Text(
                          e,
                          style: TextStyle(
                              fontSize: 14, color: Colors.grey, height: 1.6),
                        ))
                    .toList(),
              )),
              Container(
                height: 50,
                child: Row(
                  children: [
                    info.force != true
                        ? Expanded(
                            child: TextButton(
                                onPressed: () => Navigator.pop(context),
                                child: Text('取消')),
                          )
                        : Container(),
                    Expanded(
                      child: TextButton(onPressed: onOk, child: Text('升级')),
                    ),
                  ],
                ),
              )
            ],
          ),
        ),
      ),
      color: Colors.transparent,
    );
  }

  @override
  onCancel() {
    // TODO: implement onCancel
  }

  @override
  onOk() {
    AppUpgrade.upgrade(info.apkDownloadUrl);
  }
}


android/src/main/kotlin/com/wbdear/app_upgrade/app_upgrade/AppUpgradePlugin.kt
package com.wbdear.app_upgrade.app_upgrade

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Build
import androidx.annotation.NonNull
import androidx.core.content.FileProvider

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry
import java.io.File

/** AppUpgradePlugin */
class AppUpgradePlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private lateinit var channel: MethodChannel

    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "app_upgrade")
        channel.setMethodCallHandler(this)
    }

    companion object {
        @SuppressLint("StaticFieldLeak")
        lateinit var mContext: Context

        @JvmStatic
        fun registerWith(registrar: PluginRegistry.Registrar) {
            this.mContext = registrar.activity()
            val channel = MethodChannel(registrar.messenger(), "app_upgrade")
            channel.setMethodCallHandler(AppUpgradePlugin())
        }
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        mContext = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        TODO("Not yet implemented")
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        TODO("Not yet implemented")
    }

    override fun onDetachedFromActivity() {
        TODO("Not yet implemented")
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {

        when (call.method) {
            "path" -> {
                result.success(mContext.getExternalFilesDir("")?.absolutePath);
            }
            "install" -> {
                //安装app
                val path = call.argument<String>("path")
                path?.also {
                    startInstall(mContext, it)
                }
            }
            else -> {
                result.notImplemented()
            }
        }
    }


    /**
     * 安装app，android 7.0及以上和以下方式不同
     */
    private fun startInstall(context: Context, path: String) {
        val file = File(path)
        if (!file.exists()) {
            return
        }

        val intent = Intent(Intent.ACTION_VIEW)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            //7.0及以上

            val contentUri = FileProvider.getUriForFile(context, "${context.packageName}.fileprovider", file)
            intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
            intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
            intent.setDataAndType(contentUri, "application/vnd.android.package-archive")
            context.startActivity(intent)
        } else {
            //7.0以下
            intent.setDataAndType(Uri.fromFile(file), "application/vnd.android.package-archive")
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            context.startActivity(intent)
        }

    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {

    }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:sr_upgrade/sr_upgrade.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  @override
  void initState() {
    super.initState();
  }

  // Platform messages are asynchronous, so we initialize in an async method.

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Home(),
    );
  }
}

class Home extends StatefulWidget {
  @override
  _HomeState createState() => _HomeState();
}

class _HomeState extends State<Home> {
  String step = '0.00';

  _upgrade() {
    SrUpgrade.upgrade(
        'https://dldir1.qq.com/weixin/android/weixin806android1900_arm64.apk',
        progressCallback: (int count, int total) {
      setState(() {
        step = (count / total * 100).toStringAsFixed(2);
      });
    });
  }


  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Plugin example app'),
      ),
      body: Container(
        width: double.infinity,
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          crossAxisAlignment: CrossAxisAlignment.center,
          children: [
            TextButton(
              onPressed: _upgrade,
              child: Container(
                width: 200,
                height: 50,
                decoration: BoxDecoration(
                    color: Colors.blue,
                    borderRadius: BorderRadius.circular(15)),
                alignment: Alignment.center,
                child: Text(
                  '检查升级',
                  style: TextStyle(color: Colors.white, fontSize: 20),
                ),
              ),
            ),
            Text('当前下载进度$step%')
          ],
        ),
      ),
    );
  }
}
